Texture3D

tsherif opened the pull request:
- Fix 3D texture class
- Create 3D texture example, ported from PicoGL.js
- Update docs

Haven't integrated the example into the website as I'm having trouble running the website. 

For comparison, it runs quite a bit slower in Luma than in Pico on my MacBook:
- GPU render time: 17.9ms (Luma.gl) vs 14.2ms (PicoGL)
- CPU render time: 250&mu;s (Luma.gl) vs 50&mu;s (PicoGL)

I suspect the difference might be in PicoGL's state caching. It tracks the current GL bindings to avoid making redundant calls. Could be something to look into for Luma.

ibgreen commented on docs/api-reference/webgl/texture-3d.md:
```diff
@@ -1,15 +1,13 @@
 # Texture3D (WebGL2)
 
-A `Texture3D` holds a number of textures of the same size and format. The entire array can be passed to the shader which uses an extra texture coordinate to sample from it. A core feature of `Texture3D` is that the entire stack of images can passed as a single uniform to and accessed in a GLSL shader, and sampled using 3D coordinates.
```
Why cut this? These blurbs were added to all luma.gl doc pages to help new users understand what these webgl objects are about. 

Not that many people know how WebGL and especially WebGL2 classes work.

This could be worth discussing before cutting the rest of the corresponding docs

tsherif replied:
Those docs were for texture arrays. They don't apply to 3D textures.

ibgreen replied:
> Those docs were for texture arrays. They don't apply to 3D textures.

Then correct them instead of cut?

When I read the API specs, 3d textures and texture2darrays seemed very similar in that they both basically are arrays of textures and the various textures are accessed by depth, so I felt that the description here was essentially correct.

My initial challenge was to actually to understand what was different between the two, therefore, it could be helpful to users to include more insightful information about differences. Googling [here](https://stackoverflow.com/questions/35818027/understanding-the-difference-between-a-2d-texture-array-and-a-3d-texture) these differences are listed:

* GenerateMipmap will blend cross layers with 3D textures but not 2D arrays
* the Z coordinate passed to texture in GLSL is 0 to 1 with 3D textures but an 0 to N (depth) in 2D arrays.
* If filtering is not NEAREST 3D will blend across layers, 2D array will not.
* In a 3D texture, the width, height, and depth all decrease at lower mipmap sizes. In a 2D array texture, every mipmap level has the same number of array layers; only width and height decrease.

I think it would be neat to add some version of that list to the docs...

ibgreen commented on docs/api-reference/webgl/texture-3d.md:
```diff
@@ -21,15 +19,15 @@ if (Texture3D.isSupported()) {
-* `height` - height of the layer textures
-* `format` - format of the layer textures
+* `width` - width of texture
+* `height` - height of textire
```
typo: `textire`.

Again why cut: You don't feel it is valuable to emphasize that this applies to the all the texture layers to help the user understand how this object works?

tsherif replied:
Same as above. The old descriptions are for texture arrays and don't make sense for 3D texture.

ibgreen commented on examples/core/3d-texture/app.js:
```diff
@@ -0,0 +1,179 @@
+    const statsWidget = new StatsWidget(this.stats, {
+      containerStyle: 'position: absolute;top: 20px;left: 20px;'
+    });
+    statsWidget.setFormatter('CPU Time', stat => `CPU Time: ${stat.getAverageTime().toFixed(2)}ms`);
```
Are we making these formatters defaults and moving them back into the widget? I feel this adds too much code to short demo apps...

tsherif replied:
Yes, once I get this update out, I can clean up all the stats tracking code: https://github.com/uber-web/probe.gl/pull/64

ibgreen commented on modules/core/src/webgl/classes/texture-3d.js:
```diff
@@ -1,87 +1,18 @@
-  // WebGL2
-
-  // Image 3D copies from Typed Array or WebGLBuffer
-  setImage3D({
```
Nit: `setImage3D()`: I supposed is also used by e.g. Texture2DArray? Though if this is only used by 3d textures then it could be better to keep it here as it will otherwise be bundled for all apps, even if they never use a 3D texture.

ibgreen commented on modules/core/src/webgl/classes/texture.js:
```diff
@@ -489,47 +499,53 @@ export default class Texture extends Resource {
+        );
+      }
+
+      if (data instanceof Buffer) {
```
Side note: In some places I also support a `WebGLBuffer` being passed directly, as a "hack" on can do this:

```
data.handle || data
```

tsherif replied:
Do you mean doing something like this?
```js
data = data.handle || data;
if (data instanceOf WebGLBuffer) {
   //...
}
```

ibgreen replied:
Yes that works.

It does raise some API design questions about if we should document that and how consistently we can apply it around the code base. And if only for buffers or also for other objects.

Right now this probably works (for buffers) in some places and not in others.

ibgreen commented on modules/core/test/webgl/classes/index.js:
```diff
@@ -31,6 +31,7 @@ import './uniform-buffer-layout.spec';
 import './transform-feedback.spec';
 
 /*
+Spec runner can't create a WebGL 2 context
```
We do test classes with both webgl1 and webgl2 contexts, when run in browser (live/headless). In headless-gl we only test webgl1 contexts of course.

As you can see above webgl2 classes like transform-feedback etc all have test cases.

tsherif replied:
Are the browser tests still run with yarn test? I remember it used to open a browser window, but it stopped doing that recently.

tsherif replied:
Ah, my mistake. I was just looking at the node test results.

tsherif commented:
> Did you swap those numbers? GPU numbers are usually in microseconds...

No this demo is GPU intensive. It's rendering around 2 million points. CPU work is minimal.

> difference is very likely due to the fact that luma.gl examples render in retina mode by default.

I disabled that for the test. They were rendering at the same resolution (800x600).

> it should defeat any repeated calls on global settings

The unnecessary GL calls I saw (using SpectorJS) were: `useProgram` (x 2), `bindVertexArray`, `activeTexture`, `bindTexture`

tsherif commented:
I added more detail in the docs and fixed the 3d texture spec. I'm going to leave that Buffer usage for data out for now, since I want see if an how we'll actually use buffers as texture input.
